### DISCUSSION TITLE
Showing off my library for a simple binary format

### DISCUSSION THREAD

**Discussion by DJDuque** [winnow-rs/winnow#402] [Show and tell]
Winnow is amazing!

I wanted to show off [my library](https://crates.io/crates/midasio) here in case someone wants to have a look at some examples parsing a very simple binary format. The main things I think would be helpful to look at if you are doing something similar:

1.  The [seq!](https://docs.rs/winnow/0.5.30/winnow/macro.seq.html) macro is the very cool (at least for parsing simple binary formats). Some examples in my library [here](https://github.com/MIDAS-rs/midasio/blob/main/src/event.rs#L115-L139), [here](https://github.com/MIDAS-rs/midasio/blob/main/src/file.rs#L103-L130), and [here](https://github.com/MIDAS-rs/midasio/blob/main/src/data_bank.rs#L285-L296). 
2. I struggled a bit understanding how to expose an error type in my library. At the end I settled down with [something like this](https://github.com/MIDAS-rs/midasio/blob/main/src/file.rs#L15-L53) for all my structures. There is most likely a better way of doing things, but this was good enough for me to get started.
3. I also got a bit confused when I needed to write a function that "looks" into a slice of bytes instead of parsing to a structure e.g. [this](https://github.com/MIDAS-rs/midasio/blob/main/src/file.rs#L229-L243).

Overall my experience was very positive. My major issue in order to get started was probably figuring out how to keep winnow under the hood and expose e.g. error types and parsing methods in my public API.

**Comment by epage** [comment]
> The [seq!](https://docs.rs/winnow/0.5.30/winnow/macro.seq.html) macro is the very cool (at least for parsing simple binary formats). Some examples in my library [here](https://github.com/MIDAS-rs/midasio/blob/main/src/event.rs#L115-L139), [here](https://github.com/MIDAS-rs/midasio/blob/main/src/file.rs#L103-L130), and [here](https://github.com/MIDAS-rs/midasio/blob/main/src/data_bank.rs#L285-L296).

Wow, brand new and already to heavy use!

Seeing the complexity of some of them inspired me to open #403

**Comment by epage** [comment]
> I struggled a bit understanding how to expose an error type in my library. At the end I settled down with [something like this](https://github.com/MIDAS-rs/midasio/blob/main/src/file.rs#L15-L53) for all my structures. There is most likely a better way of doing things, but this was good enough for me to get started.

This is roughly what I would expect.  For myself, I generally have few entry points into my parser, so I would have just had a private constructor for my public error type that takes `ParseError` and put that in `map_err` calls, rather than defining an inner error type.

**Comment by epage** [comment]
> I also got a bit confused when I needed to write a function that "looks" into a slice of bytes instead of parsing to a structure e.g. [this](https://github.com/MIDAS-rs/midasio/blob/main/src/file.rs#L229-L243).

If I'm understanding this, you are looking for something more like a regex search but `Parser::parse` requires you to handle the whole content.  Is that right?

One alternative is to use `Parser::parse_next` though that requires to you still get the error into the shape you need for your general conversions.

> `take_while(0.., |_| true),`

You could use `rest` for this, right?
